### PR TITLE
spec out new extension api

### PR DESCRIPTION
**Pull request by mafintosh**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -429,9 +428,27 @@ The returned object is of the form:
...
+{
+  encoding: 'json' | 'binary' | 'utf-8' | anyAbstractEncoding
+  onmessage (message, peer) {
+    // called when a message is received from a peer
+    // will be decoded using the encoding you provide
+  }
+}
+```
+
+#### `ext.send(message, peer)`
```
Comment (andrewosh) [inline README.md]: Is the `peer` param a `Peer` object or is it an ID? Maybe describe that `peer` is the object emitted on the `peer` event.

File: index.js [hunk 2]
```diff
@@ -208,13 +206,90 @@ Feed.prototype.replicate = function (initiator, opts) {
...
+    if (!encoding) {
+      this.handlers.onmessage(buffer, peer)
+      return
+    }
+
+    let message = null
+
+    try {
+      message = encoding.decode(buffer)
+    } catch (_) {
```
Comment (andrewosh) [inline index.js]: Should extension errors be propagated?

File: lib/replicate.js [hunk 3]
```diff
@@ -46,6 +46,8 @@ function replicate (feed, initiator, opts) {
 }
 
 function Peer (feed, opts) {
+  if (opts.extensions) throw new Error('Per peer extensions is deprecated. Use feed.registerExtension instead')
```
Comment (andrewosh) [inline lib/replicate.js]: If it's just deprecated it shouldn't throw an error. Update error message to say "not supported"

File: lib/replicate.js [hunk 4]
```diff
@@ -201,6 +202,15 @@ Peer.prototype.onrequest = function (request) {
   }
 }
 
+Peer.prototype._updateOptions = function () {
+  if (this.ack || (this.feed.extensions.length)) {
```
Comment (andrewosh) [inline lib/replicate.js]: What does an options update have to do with the `ack` flag?

Comment (mafintosh) [reply lib/replicate.js]: It needs to send an options message if ack is enabled OR you are using extensions, othersise it is not needed

Comment (okdistribute) [reply lib/replicate.js]: Seems like ack could be implemented as an extension, although might be more work than it's worth

Comment (mafintosh) [reply lib/replicate.js]: Yea, we should consider that down the line I think

**Pull request merged**